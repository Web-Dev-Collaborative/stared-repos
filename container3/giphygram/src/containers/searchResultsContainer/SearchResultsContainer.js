import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { SearchResults } from '../../components/searchResults/SearchResults';
import { InfiniteScroll } from '../../components/shared/infiniteScroll/InfiniteScroll';
import { Spinner } from '../../components/shared/spinner/Spinner';
import { searchMore } from '../../actions/searchActions';
import { getSearchResultsFromState } from '../../reducers/searchResultsReducer';
import { getLayoutFromState } from '../../reducers/layoutReducer';
import { DEFAULT_COLUMNS_NUM, LayoutService, SUPPORTED_COLUMNS_NUMS } from '../../services/LayoutService';

export class SearchResultsContainer extends React.Component {
  static propTypes = {
    searchItems: PropTypes.arrayOf(PropTypes.object),
    isLoading: PropTypes.bool,
    isFetchingMore: PropTypes.bool,
    searchMore: PropTypes.func.isRequired,
    columnsNum: PropTypes.oneOf(SUPPORTED_COLUMNS_NUMS),
  };

  static defaultProps = {
    searchItems: [],
    isLoading: false,
    isFetchingMore: false,
    columnsNum: DEFAULT_COLUMNS_NUM,
  };

  onFetchMore = () => {
    const { isFetchingMore, isLoading, searchMore: searchMoreCallback } = this.props;

    if (!isFetchingMore && !isLoading) {
      searchMoreCallback();
    }
  };

  render() {
    const {
      searchItems, isLoading, isFetchingMore, columnsNum,
    } = this.props;

    const fetchMoreSpinner = isFetchingMore ? <Spinner /> : null;

    return (
      <InfiniteScroll onFetchMore={this.onFetchMore}>
        <SearchResults
          searchItems={searchItems}
          isLoading={isLoading}
          columnsNum={columnsNum}
        />
        {fetchMoreSpinner}
      </InfiniteScroll>
    );
  }
}

const mapStateToProps = (state) => {
  const searchResults = getSearchResultsFromState(state);
  const layout = getLayoutFromState(state);

  // Decide how many columns to display based on responsive width and on user selection.
  const responsiveColumnsNum = LayoutService.screenSizeToColumns(layout.size);
  const columnsNum = LayoutService.choseColumnsNum(
    responsiveColumnsNum,
    layout.manualColumnsNum,
  );

  return {
    searchItems: searchResults.data || [],
    isLoading: searchResults.isLoading || false,
    isFetchingMore: searchResults.isFetchingMore || false,
    columnsNum,
  };
};

const mapDispatchToProps = {
  searchMore,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchResultsContainer);
