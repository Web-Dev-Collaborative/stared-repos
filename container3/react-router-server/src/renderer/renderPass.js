import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
