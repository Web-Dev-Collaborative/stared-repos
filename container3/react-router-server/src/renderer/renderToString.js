import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
