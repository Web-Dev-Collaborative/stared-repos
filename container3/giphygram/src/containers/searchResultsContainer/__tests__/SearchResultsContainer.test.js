import React from 'react';
import renderer from 'react-test-renderer';
import { SearchResultsContainer } from '../SearchResultsContainer';

jest.mock(
  '../../../components/shared/infiniteScroll/InfiniteScroll',
  () => ({
    InfiniteScroll: 'InfiniteScroll',
  }),
);

jest.mock(
  '../../../components/searchResults/SearchResults',
  () => ({
    SearchResults: 'SearchResults',
  }),
);

describe('SearchResultsContainer', () => {
  it('should be rendered correctly', () => {
    const searchMore = jest.fn();

    const testElement = renderer.create((
      <SearchResultsContainer searchMore={searchMore} />
    ));

    expect(testElement.toJSON()).toMatchSnapshot();
    expect(searchMore).not.toHaveBeenCalled();
  });

  it('should be rendered correctly with custom columns num', () => {
    const searchMore = jest.fn();

    const testElement = renderer.create((
      <SearchResultsContainer searchMore={searchMore} columnsNum={4} />
    ));

    expect(testElement.toJSON()).toMatchSnapshot();
    expect(searchMore).not.toHaveBeenCalled();
  });

  it('should fire searchMore callback', () => {
    const searchMore = jest.fn();

    const componentInstance = renderer.create((
      <SearchResultsContainer searchMore={searchMore} />
    )).root;

    expect(searchMore).not.toHaveBeenCalled();

    const infiniteScroller = componentInstance.findByType('InfiniteScroll');
    expect(infiniteScroller).toBeDefined();

    infiniteScroller.props.onFetchMore();
    expect(searchMore).toHaveBeenCalledTimes(1);
  });

  it('should not fire searchMore callback in case if loading is in progress', () => {
    const searchMore = jest.fn();

    const componentInstance = renderer.create((
      <SearchResultsContainer searchMore={searchMore} isLoading />
    )).root;

    expect(searchMore).not.toHaveBeenCalled();

    const infiniteScroller = componentInstance.findByType('InfiniteScroll');
    expect(infiniteScroller).toBeDefined();

    infiniteScroller.props.onFetchMore();
    expect(searchMore).not.toHaveBeenCalled();
  });

  it('should not fire searchMore callback in case if fetching more is in progress', () => {
    const searchMore = jest.fn();

    const componentInstance = renderer.create((
      <SearchResultsContainer searchMore={searchMore} isFetchingMore />
    )).root;

    expect(searchMore).not.toHaveBeenCalled();

    const infiniteScroller = componentInstance.findByType('InfiniteScroll');
    expect(infiniteScroller).toBeDefined();

    infiniteScroller.props.onFetchMore();
    expect(searchMore).not.toHaveBeenCalled();
  });
});
