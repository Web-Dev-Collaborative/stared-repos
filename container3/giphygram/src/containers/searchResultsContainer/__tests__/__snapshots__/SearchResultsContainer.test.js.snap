// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsContainer should be rendered correctly 1`] = `
<InfiniteScroll
  onFetchMore={[Function]}
>
  <SearchResults
    columnsNum={3}
    isLoading={false}
    searchItems={Array []}
  />
</InfiniteScroll>
`;

exports[`SearchResultsContainer should be rendered correctly with custom columns num 1`] = `
<InfiniteScroll
  onFetchMore={[Function]}
>
  <SearchResults
    columnsNum={4}
    isLoading={false}
    searchItems={Array []}
  />
</InfiniteScroll>
`;
